### DISCUSSION TITLE
How to check whether two Coord objects have the same metadata?

### DISCUSSION THREAD

**Discussion by krikru** [SciTools/iris#5409] [Q&A]
When you call [`concatenate_cube`](https://scitools-iris.readthedocs.io/en/stable/generated/api/iris.cube.html#iris.cube.CubeList.concatenate_cube) on a [`CubeList`](https://scitools-iris.readthedocs.io/en/stable/generated/api/iris.cube.html#iris.cube.CubeList) object, the concatenation will fail and throw a [`ConcatenateError`](https://scitools-iris.readthedocs.io/en/stable/generated/api/iris.exceptions.html#iris.exceptions.ConcatenateError) exception that says that "Dimension coordinates metadata differ" if two corresponding [`Coord`](https://scitools-iris.readthedocs.io/en/stable/generated/api/iris.coords.html#iris.coords.Coord) objects have differing metadata. Is there any robust way to check beforehand whether this metadata differs between two `Coord` objects? I can't find any method that returns the metadata of such an object.

The reason I'm asking is that I have a bunch of netCDF files that I want to concatenate, and apparently some of them have slightly deviating coordinate values (they are off by something like a millionth of a degree), and others have deviating dtype (`float64` instead of `float32`), and I want to detect any discrepancy so that I can notify the user of it, and perhaps also fix it automatically. That means that I need a robust way to check not only whether two `Coord` objects are equal, but also whether their metadata is, and this comparison is not performed (at least not completely) when I just check if the two objects are equal.

**Comment by acchamber** [comment]
Not in Iris currently. Better comparison and information when comparing coords is something iris hopes to adress in the future, but for now they [cube helper](https://cube-helper.readthedocs.io/en/v2.2.3/Cube%20Helper%20Tutorial.html) package may be of some use rather than inventing the wheel yourself.

**Comment by bjlittle** [comment]
@krikru As of `iris` 3.0 we implemented the [common metadata API](https://scitools-iris.readthedocs.io/en/latest/further_topics/metadata.html#metadata) within `iris` which allows you to compare, combine and detect the differences in coordinate (and other `iris` objects) metadata.

For example, to get the metadata of an `AncillaryVariable`, `AuxCoord`, `AuxCoordFactory`, `CellMeasure`, `Cube` or `DimCoord` simply use the common `metadata` property i.e., `coord.metadata`

There is reasonably rich metadata behaviour supported that allows you to easily perform:

- [metadata equality](https://scitools-iris.readthedocs.io/en/latest/further_topics/metadata.html#metadata-equality) e.g., `coord1.metadata == coord2.metadata`
- [metadata difference](https://scitools-iris.readthedocs.io/en/latest/further_topics/metadata.html#metadata-difference) e.g., `coord1.metadata.difference(coord2.metadata)`
- [metadata combination](https://scitools-iris.readthedocs.io/en/latest/further_topics/metadata.html#metadata-combination) e.g., `coord1.metadata = coord1.metadata.combine(coord2.metadata)`
- [metadata conversion](https://scitools-iris.readthedocs.io/en/latest/further_topics/metadata.html#metadata-conversion) e.g., `coord.metadata.from_metadata(cube.metadata)`
- [metadata assignment](https://scitools-iris.readthedocs.io/en/latest/further_topics/metadata.html#metadata-assignment) provides various different convenient ways to populate and assign metadata to an `iris` instance

Note also, we have `strict` and `lenient` metadata behaviours, see [here](https://scitools-iris.readthedocs.io/en/latest/further_topics/lenient_metadata.html#lenient-metadata) for further details.

The upshot is that you can easily see the difference between two coordinates, as you asked above, but there isn't an automated way (at the moment) in which concatenate will tell you all the metadata differences.

Does that help?

**Reply by trexfeathers** [reply]
I don't want to speak for @krikru, but the `"Dimension coordinates metadata differ"` error in `concatenate()` is NOT referring to the above metadata. I recently had this problem with @HelenRJohnson and we discovered that it has its own definition of 'metadata', which includes `dtype` for instance.

#5395 will fix this

**Reply by acchamber** [reply]
In particular, concatenate compares the `dim_metadata` and `aux_metadata` objects of `_CubeSignature`  objects that are prepared during the the `concatenate() ` process.
